metals/src/main/scala/scala/meta/internal/metals/ImplementationProvider.scala

```diff
@@ -0,0 +1,121 @@
+import scala.meta.internal.semanticdb.SymbolOccurrence
+import scala.meta.internal.semanticdb.ClassSignature
+import scala.meta.internal.semanticdb.TypeRef
+import scala.meta.internal.semanticdb.Signature
+import scala.meta.internal.semanticdb.TextDocument
+import java.util.concurrent.ConcurrentHashMap
+import java.nio.file.Path
+
+final class ImplementationProvider(
+    semanticdbs: Semanticdbs,
+    workspace: AbsolutePath,
+    buffer: Buffers,
+    definitionProvider: DefinitionProvider
+) {
+  private val implementationsInPath =
+    new ConcurrentHashMap[Path, Map[String, Set[ClassLocation]]]
+
+  def implementations(params: TextDocumentPositionParams): List[Location] = {
+    val source = params.getTextDocument.getUri.toAbsolutePath
+
+    def findSemanticdb(fileSource: AbsolutePath) =
+      semanticdbs
+        .textDocument(fileSource)
+        .documentIncludingStale
+        .toList
+
+    for {
+      currentDoc <- findSemanticdb(source)
+      positionOccurrence = definitionProvider.positionOccurrence(
+        source,
+        params,
+        currentDoc
+      )
+      occ <- positionOccurrence.occurrence.toList
+      (file, locations) <- findImplementation(occ.symbol).groupBy(_.file)
+      fileSource = AbsolutePath(file)
+      doc <- findSemanticdb(fileSource)
+      distance = if (fileSource == source) positionOccurrence.distance
+      else TokenEditDistance.fromBuffer(fileSource, doc.text, buffer)
+      impl <- locations
+      range <- impl.symbol.range
+      revised <- distance.toRevised(range.toLSP)
+      uri = impl.file.toUri.toString
+    } yield new Location(uri, revised)
+  }
+
+  private def findImplementation(symbol: String): Set[ClassLocation] = {
+    def findAllImpl(symbol: String): Set[ClassLocation] = {
+      val directImpl = for {
+        (path, symbols) <- implementationsInPath.asScala
+        symbolImpls <- symbols.get(symbol).toList
+        impl <- symbolImpls
+      } yield impl
+      directImpl.toSet ++ directImpl
+        .flatMap(
+          loc => findAllImpl(loc.symbol.symbol)
+        )
+    }
+    findAllImpl(symbol)
+  }
+
+  def clear(): Unit = {
+    implementationsInPath.clear()
+  }
+
+  def onDelete(path: Path): Unit = {
+    implementationsInPath.remove(path)
+  }
+
+  def onChange(docs: TextDocuments, path: Path): Unit = {
+    implementationsInPath.compute(
+      path, { (_, _) =>
+        val allParents = for {
+          doc <- docs.documents
+          thisSymbol <- doc.symbols
+          occ <- doc.occurrences
+            .find(
+              occ => occ.symbol == thisSymbol.symbol && occ.role.isDefinition
+            )
+            .toList
+          parents <- addFromSignature(thisSymbol.signature, occ, doc).toList
+          parent <- parents
+        } yield parent
+
+        allParents.groupBy(_._1).map {
+          case (symbol, locations) =>
+            symbol -> locations.map(_._2).toSet
+        }
+      }
+    )
+  }
+
+  private def addFromSignature(
```

Comment:
it returns `Option[List[(Symbol, Location)]]` right? We could easily make it a `List[(Symbol, Location)]` which would simplify `onChange` a bit:
```scala
          parents <- addFromSignature(thisSymbol.signature, occ, doc).toList
          parent <- parents
        } yield parent
```

would become 
```scala
          parent <- addFromSignature(thisSymbol.signature, occ, doc).toList
        } yield parent
```

Reply:
Switched to Seq and it simplified the code.